cli-plugins/manager/hooks.go

```diff
@@ -0,0 +1,180 @@
+//
+// Example:
+//
+// "run this image with `docker run " + TemplateReplaceArg(0) + "`"
+//
+// when being executed after the command:
+// `docker pull alpine`
+// will result in the message:
+// "Run this image with `docker run alpine`"
+func TemplateReplaceArg(i int) string {
+	return fmt.Sprintf(hookTemplateArg, strconv.Itoa(i))
+}
+
+// RunPluginHooks calls the hook subcommand for all present
+// CLI plugins that declare support for hooks in their metadata
+// and parses/prints their responses
+func RunPluginHooks(dockerCli command.Cli, rootCmd, subCommand *cobra.Command) error {
+	plugins, err := ListPlugins(dockerCli, rootCmd)
+	if err != nil {
+		return err
+	}
+	subCmdName := subCommand.Name()
+
+	hookMessages := make([]string, 0, len(plugins))
+	for _, p := range plugins {
+		p := p
+		if !p.supportsHookCommand(subCmdName) {
+			continue
+		}
+
+		hookReturn, err := p.RunHook(subCmdName)
+		// Skip misbehaving plugins, but don't halt execution
+		if err != nil {
+			continue
+		}
+		hookMessages = append(hookMessages, parseHooks(hookReturn.Template, subCommand))
+	}
+
+	printHookMessages(dockerCli.Err(), hookMessages)
+	return nil
+}
+
+const (
+	hookTemplateCommandName = "{docker-cli-plugin-hook-cmd-name}"
+	hookTemplateFlagValue   = "{docker-cli-plugin-hook-flag-%s}"
+	hookTemplateArg         = "{docker-cli-plugin-hook-arg-%s}"
+)
+
+func parseHooks(template string, cmd *cobra.Command) string {
+	ret := template
+	for {
+		parsed, foundMatch := parseFlagValue(ret, cmd)
+		ret = parsed
+		if !foundMatch {
+			break
+		}
+	}
+	for {
+		parsed, foundMatch := parseArg(ret, cmd)
+		ret = parsed
+		if !foundMatch {
+			break
+		}
+	}
+	return strings.ReplaceAll(ret, hookTemplateCommandName, cmd.Name())
+}
+
+var (
+	flagRegex = regexp.MustCompile(fmt.Sprintf(hookTemplateFlagValue, `([\w-]+)`))
+	argRegex  = regexp.MustCompile(fmt.Sprintf(hookTemplateArg, `([\w-]+)`))
+)
+
+func parseFlagValue(template string, cmd *cobra.Command) (string, bool) {
+	matches := flagRegex.FindStringSubmatch(template)
+	if len(matches) == 0 {
+		return template, false
+	}
+	for _, m := range matches[1:] {
+		flag := cmd.Flag(m)
+		if flag == nil {
+			continue
+		}
+		value := flag.Value.String()
+		template = strings.ReplaceAll(template, TemplateReplaceFlagValue(m), value)
+	}
+	return template, true
+}
+
+func parseArg(template string, cmd *cobra.Command) (string, bool) {
+	matches := argRegex.FindStringSubmatch(template)
+	if len(matches) == 0 {
+		return template, false
+	}
+	for _, m := range matches[1:] {
+		i, err := strconv.Atoi(m)
+		if err != nil {
+			continue
+		}
+		if i >= len(cmd.Flags().Args()) {
+			return template, false
+		}
+		template = strings.ReplaceAll(template, TemplateReplaceArg(i), cmd.Flags().Arg(i))
+	}
+	return template, true
+}
+
+func printHookMessages(out io.Writer, messages []string) {
+	if len(messages) == 0 {
+		return
+	}
+
+	fmt.Fprintln(out, aec.LightBlueF.Apply("\nNext steps:"))
```

Comment:
I like the idea of having different message types:
https://github.com/docker/cli/blob/10e34442b14df1fe57f689673f4f98901e49ae24/cli-plugins/manager/hooks.go#L27-L28

I think it makes sense to implement this now, with only one kind `NextSteps`, so that it's clear the the CLI is "the host" here and specifies the layout. If some plugin needs a different kind of layout, the CLI needs to be aware of that.

WDYT?

Reply:
This could be nice to have, especially once we start working more on the cli ux itself. Not sure if we should consider the enum to represent the "type" of hook, but surely it'd be good to have some differentiation for ux purposes

Reply:
Yeah, I'm of that mind as well. I'll update the PR to remove the TODO and codify this in.